File: pkg/heartbeat/heartbeat.go
package heartbeat

// Metrics represents the normalized metrics sent from agent to SaaS.
// Agent must populate these fields with consistent units.
type Metrics struct {
	CPUPercent  float64 `json:"cpu_percent"`
	MemoryBytes uint64  `json:"memory_bytes"` // RSS in bytes
	Goroutines  int     `json:"goroutines"`
	NetRxBytes  uint64  `json:"net_rx_bytes"`
	NetTxBytes  uint64  `json:"net_tx_bytes"`
	UptimeS     uint64  `json:"uptime_s"`
}

// Request is the heartbeat payload sent by the agent.
type Request struct {
	AgentID          string  `json:"agent_id"`
	SessionID        string  `json:"session_id"`
	UptimeS          uint64  `json:"uptime_s"`
	ActiveExecutions int     `json:"active_executions"`
	Metrics          Metrics `json:"metrics"`
}

File: web/src/types/heartbeat.ts
export interface HeartbeatMetrics {
  organization_id: string
  agent_id: string
  session_id: string
  uptime_s: number
  active_executions: number
  cpu_percent: number
  memory_bytes: number
  net_rx_bytes: number
  net_tx_bytes: number
}

export interface MetricsPoint {
  timestamp: string
  cpu_percent: number
  memory_bytes: number
  net_rx_bytes: number
  net_tx_bytes: number
}
